// stardust-core/dist/common/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/** A dictionary class that maps string to ValueType */
var Dictionary = /** @class */ (function () {
    function Dictionary() {
        this._dict = {};
    }
    /** Set an entry */
    Dictionary.prototype.set = function (key, value) {
        this._dict[key] = value;
    };
    /** Determine if the dictionary has an entry */
    Dictionary.prototype.has = function (key) {
        return this._dict.hasOwnProperty(key);
    };
    /** Delete an entry from the dictionary */
    Dictionary.prototype.delete = function (key) {
        delete this._dict[key];
    };
    /** Get a entry, if not found, return undefined */
    Dictionary.prototype.get = function (key) {
        if (this._dict.hasOwnProperty(key)) {
            return this._dict[key];
        }
        else {
            return undefined;
        }
    };
    /** Iterate over the dictionary */
    Dictionary.prototype.forEach = function (cb) {
        for (var key in this._dict) {
            if (this._dict.hasOwnProperty(key)) {
                cb(this._dict[key], key);
            }
        }
    };
    /** Create a copy of the dictionary */
    Dictionary.prototype.clone = function () {
        var result = new Dictionary();
        this.forEach(function (value, key) {
            result.set(key, value);
        });
        return result;
    };
    return Dictionary;
}());
exports.Dictionary = Dictionary;
/** Shallow clone an object */
function shallowClone(object) {
    var result = {};
    for (var key in object) {
        if (object.hasOwnProperty(key)) {
            result[key] = object[key];
        }
    }
    return result;
}
exports.shallowClone = shallowClone;
/** Attempt different names starting with prefix until check returns true */
function attemptName(prefix, check) {
    if (check(prefix)) {
        return prefix;
    }
    for (var i = 1;; i++) {
        var c = prefix + i.toString();
        if (check(c)) {
            return c;
        }
    }
}
exports.attemptName = attemptName;
