// stardust-core/dist/core/exceptions.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
// Base class for errors.
var BaseError = (function (_super) {
    __extends(BaseError, _super);
    function BaseError(message) {
        _super.call(this, message);
        this.stack = (new Error(message)).stack;
    }
    return BaseError;
}(Error));
exports.BaseError = BaseError;
// Parse error.
var ParseError = (function (_super) {
    __extends(ParseError, _super);
    function ParseError(message, start, end) {
        _super.call(this, message);
        this.name = "ParseError";
        this.message = message;
        this.start = start;
        this.end = end;
    }
    return ParseError;
}(BaseError));
exports.ParseError = ParseError;
// Compile error.
var CompileError = (function (_super) {
    __extends(CompileError, _super);
    function CompileError(message, start, end) {
        _super.call(this, message);
        this.name = "CompileError";
        this.message = message;
        this.start = start;
        this.end = end;
    }
    return CompileError;
}(BaseError));
exports.CompileError = CompileError;
// Runtime error.
var RuntimeError = (function (_super) {
    __extends(RuntimeError, _super);
    function RuntimeError(message, start, end) {
        _super.call(this, message);
        this.name = "RuntimeError";
        this.message = message;
        this.start = start;
        this.end = end;
    }
    return RuntimeError;
}(BaseError));
exports.RuntimeError = RuntimeError;
