// stardust-core/dist/core/platform/platform.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var utils_1 = require("../utils/utils");
var PlatformMarkData = (function () {
    function PlatformMarkData() {
    }
    return PlatformMarkData;
}());
exports.PlatformMarkData = PlatformMarkData;
var PlatformMark = (function () {
    function PlatformMark() {
    }
    return PlatformMark;
}());
exports.PlatformMark = PlatformMark;
var Viewport = (function () {
    function Viewport() {
    }
    return Viewport;
}());
exports.Viewport = Viewport;
var Viewport2D = (function (_super) {
    __extends(Viewport2D, _super);
    function Viewport2D(width, height) {
        _super.call(this);
        this.width = width;
        this.height = height;
    }
    return Viewport2D;
}(Viewport));
exports.Viewport2D = Viewport2D;
var Viewport3D = (function (_super) {
    __extends(Viewport3D, _super);
    function Viewport3D(width, height, fov) {
        _super.call(this);
        this.width = width;
        this.height = height;
        this.fov = fov;
    }
    return Viewport3D;
}(Viewport));
exports.Viewport3D = Viewport3D;
var Platform = (function () {
    function Platform() {
    }
    return Platform;
}());
exports.Platform = Platform;
var platformConstructors = new utils_1.Dictionary();
function registerPlatformConstructor(name, ctor) {
    platformConstructors.set(name, ctor);
}
exports.registerPlatformConstructor = registerPlatformConstructor;
function platform(name) {
    var args = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        args[_i - 1] = arguments[_i];
    }
    return platformConstructors.get(name).apply(void 0, args);
}
exports.platform = platform;
