// stardust-core/dist/binding/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var common_1 = require("../common");
var BindingType;
(function (BindingType) {
    BindingType[BindingType["VALUE"] = 0] = "VALUE";
    BindingType[BindingType["FUNCTION"] = 1] = "FUNCTION";
    BindingType[BindingType["TEXTURE"] = 2] = "TEXTURE";
})(BindingType = exports.BindingType || (exports.BindingType = {}));
var ShiftBinding = /** @class */ (function () {
    function ShiftBinding(name, offset) {
        this.name = name;
        this.offset = offset;
    }
    return ShiftBinding;
}());
exports.ShiftBinding = ShiftBinding;
// Resolve binding primitives to Value (Value = number or number[]).
function getBindingValue(value) {
    if (value instanceof common_1.MathType) {
        return value.toArray();
    }
    else {
        return value;
    }
}
exports.getBindingValue = getBindingValue;
exports.types = {
    float: { name: "float", size: 4, primitive: "float", primitiveCount: 1 },
    int: { name: "int", size: 4, primitive: "int", primitiveCount: 1 },
    Vector2: { name: "Vector2", size: 8, primitive: "float", primitiveCount: 2 },
    Vector3: { name: "Vector3", size: 12, primitive: "float", primitiveCount: 3 },
    Quaternion: {
        name: "Quaternion",
        size: 16,
        primitive: "float",
        primitiveCount: 4
    },
    Color: { name: "Color", size: 16, primitive: "float", primitiveCount: 4 }
};
